pkg/raft/term_cache_test.go

```diff
@@ -0,0 +1,343 @@
+			// Initial: [t5/10, t6/11, t7/12, t8/13]
+			init: entryID{term: 5, index: 10}.append(6, 7, 8).LogSlice,
+			// Append: [t5/10, t7/11, t8/12, t9/13, t10/14, t11/15, t12/16, t13/17, t13/18, ...]
+			app: entryID{term: 5, index: 10}.append(7, 8, 9, 10, 11, 12, 13, 13, 13, 13, 13, 13).LogSlice,
+			// Cached: [t10/14, t11/15, t12/16, t13/17, t13/18, ...]
+			want: entryID{term: 0, index: 13}.append(10, 11, 12, 13, 13, 13, 13, 13, 13).LogSlice,
+		},
+		{
+			// Extreme but possible, overwrite with LogSlice.prev and entry.
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11, t7/12, t8/13]
+			init: entryID{term: 5, index: 10}.append(6, 7, 8).LogSlice,
+			// Append: [t4/9, t10/10]
+			app: entryID{term: 4, index: 9}.append(10).LogSlice,
+			// Cached: [t4/9, t10/10]
+			want: entryID{term: 0, index: 8}.append(4, 10).LogSlice,
+		},
+		{
+			// Overwrite with just one entry.
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11, t7/12, t8/13]
+			init: entryID{term: 5, index: 10}.append(6, 7, 8).LogSlice,
+			// Append: [t5/10, t10/11]
+			app: entryID{term: 5, index: 10}.append(10).LogSlice,
+			// Cached: [t5/10, t10/11]
+			want: entryID{term: 0, index: 9}.append(5, 10).LogSlice,
+		},
+		{
+			// Multiple gaps in between, test term.
+			cacheSize: 4,
+			// Initial: [t5/10, t6/11]
+			init: entryID{term: 5, index: 10}.append(6).LogSlice,
+			// Append: [t7/12, t7/13, t7/14, t9/15, t9/16, t9/17, t10/18, t10/19, t10/20]
+			app: entryID{term: 6, index: 11}.append(7, 7, 7, 9, 9, 9, 10, 10, 10).LogSlice,
+			// Cached: [t6/11, t7/12, t7/13, t7/14, t9/15, t9/16, t9/17, t10/18, t10/19, t10/20]
+			want: entryID{term: 0, index: 10}.append(6, 7, 7, 7, 9, 9, 9, 10, 10, 10).LogSlice,
+		},
+	} {
+		t.Run("", func(t *testing.T) {
+			// Initialize cache with the first entry of init
+			cache := NewTermCache(tt.cacheSize, tt.init.prev)
+			cache.checkInvariants(t)
+			// First append to establish initial state
+			cache.truncateAndAppend(tt.init)
+			cache.checkInvariants(t)
+			// Second append with new entries
+			cache.truncateAndAppend(tt.app)
+			cache.checkInvariants(t)
+
+			// tt.want.prev.index represents one index below what is covered by cache
+			// Anything at or below prev should also return found == false
+			_, found := cache.term(tt.want.prev.index)
+			require.False(t, found, "Looking up for term below the cache should result in not found")
+
+			for _, expTermAtIdx := range tt.want.entries {
+				actualTerm, found := cache.term(expTermAtIdx.Index)
+				require.True(t, found, "Should find term at index %d", expTermAtIdx.Index)
+				require.Equal(t, expTermAtIdx.Term, actualTerm, "unexpected term for index %d", expTermAtIdx.Index)
+			}
+
+			// Explicit check that looking up a few indices above the last entry appended
+			// to cache returns the same term as the last appended entry's term.
+			// Assuming want.last to be the last entryID appended to cache.
+			last := len(tt.want.entries) - 1
+			for i := tt.want.entries[last].Index; i < 3; i++ {
```

Comment:
This looks like a no-op if the last index is >= 3. You probably intended to increment it a couple of times unconditionally.